File: tests/test_fastscan_filter.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

import unittest

import numpy as np
import faiss

from common_faiss_tests import for_all_simd_levels
from faiss.contrib import datasets

faiss.omp_set_num_threads(4)


@for_all_simd_levels
class TestFastScanFiltering(unittest.TestCase):
    """
    Test IDSelector filtering on IVF fast_scan indexes.

    The PR adds block-skip filtering to the fast_scan accumulate loops:
    before processing each 32-vector block, it checks if all vectors
    in the block are filtered out by the IDSelector, and skips the
    block entirely if so.

    These tests verify that filtered results only contain allowed IDs.
    Only IVF-based fastscan indexes support SearchParameters with sel.
    """

    def do_test_filter(
        self,
        index_key,
        id_selector_type="batch",
        mt=faiss.METRIC_L2,
        k=10,
        nb=1000,
    ):
        """
        Build a fastscan IVF index and search with IDSelector filtering.
        Verify all returned results are in the allowed set.
        """
        d = 32
        ds = datasets.SyntheticDataset(d, 2000, nb, 20, metric=mt)

        index = faiss.index_factory(d, index_key, mt)
        index.train(ds.get_train())
        index.add(ds.get_database())

        # Create selector and allowed set
        rs = np.random.RandomState(123)
        if id_selector_type == "batch":
            subset = rs.choice(nb, nb // 3, replace=False).astype("int64")
            sel = faiss.IDSelectorBatch(subset)
            allowed = set(subset.tolist())
        elif id_selector_type == "range":
            lo, hi = nb // 4, 3 * nb // 4
            sel = faiss.IDSelectorRange(lo, hi)
            allowed = set(range(lo, hi))
        elif id_selector_type == "not_batch":
            # Exclude entire 32-vector blocks to test block-skip
            excluded = np.concatenate([
                np.arange(0, 32, dtype="int64"),
                np.arange(64, 96, dtype="int64"),
            ])
            inner_sel = faiss.IDSelectorBatch(excluded)
            sel = faiss.IDSelectorNot(inner_sel)
            allowed = set(i for i in range(nb) if i not in excluded)
        elif id_selector_type == "partial_batch":
            # Exclude a few IDs within a single block (not a whole block)
            excluded = np.array([5, 10, 20, 31], dtype="int64")
            inner_sel = faiss.IDSelectorBatch(excluded)
            sel = faiss.IDSelectorNot(inner_sel)
            allowed = set(i for i in range(nb) if i not in excluded)
        elif id_selector_type == "empty":
            sel = faiss.IDSelectorBatch(np.array([], dtype="int64"))
            allowed = set()
        else:
            raise ValueError(f"Unknown selector type: {id_selector_type}")

        params = faiss.SearchParametersIVF(sel=sel, nprobe=8)
        Dfs, Ifs = index.search(ds.get_queries(), k, params=params)

        # Check: all results are in the allowed set or -1
        for q in range(ds.nq):
            for j in range(k):
                idx = int(Ifs[q, j])
                if idx >= 0:
                    self.assertIn(
                        idx, allowed,
                        f"Query {q}, rank {j}: got id {idx} not in allowed set"
                    )

        # If allowed set is large enough, expect some valid results
        if len(allowed) > k:
            valid = np.sum(Ifs >= 0)
            self.assertGreater(
                valid, 0, "Expected some valid results"
            )

        return Ifs

    # ------- IVFPQFastScan tests -------

    def test_IVFPQfs_batch_L2(self):
        self.do_test_filter("IVF32,PQ4x4fs", "batch", faiss.METRIC_L2)

    def test_IVFPQfs_batch_IP(self):
        self.do_test_filter(
            "IVF32,PQ4x4fs", "batch", faiss.METRIC_INNER_PRODUCT
        )

    def test_IVFPQfs_range(self):
        self.do_test_filter("IVF32,PQ4x4fs", "range")

    def test_IVFPQfs_not_batch(self):
        """IDSelectorNot excluding whole blocks -> block-skip path"""
        self.do_test_filter("IVF32,PQ4x4fs", "not_batch")

    def test_IVFPQfs_partial_batch(self):
        """IDSelectorNot excluding a few IDs within a block"""
        self.do_test_filter("IVF32,PQ4x4fs", "partial_batch")

    def test_IVFPQfs_empty_selector(self):
        """Empty selector accepts nothing -> all results -1"""
        Ifs = self.do_test_filter("IVF32,PQ4x4fs", "empty")
        np.testing.assert_array_equal(Ifs, -1)

    # ------- Different k values (different handler paths) -------

    def test_IVFPQfs_k1(self):
        """k=1 -> SingleResultHandler path"""
        self.do_test_filter("IVF32,PQ4x4fs", "batch", k=1)

    def test_IVFPQfs_k40(self):
        """k=40 -> ReservoirHandler path"""
        self.do_test_filter("IVF32,PQ4x4fs", "batch", k=40)

    # ------- Non-aligned ntotal -------

    def test_IVFPQfs_ntotal_50(self):
        """ntotal=50, not a multiple of 32"""
        self.do_test_filter("IVF32,PQ4x4fs", "batch", nb=50)

    def test_IVFPQfs_ntotal_77(self):
        """ntotal=77, partial last block"""
        self.do_test_filter("IVF32,PQ4x4fs", "batch", nb=77)

    def test_IVFPQfs_ntotal_150(self):
        """ntotal=150, not a multiple of 32"""
        self.do_test_filter("IVF32,PQ4x4fs", "batch", nb=150)


@for_all_simd_levels
class TestBlockSkipConsistency(unittest.TestCase):
    """
    Test that block-skip filtering produces consistent results
    across different filter configurations on fastscan IVF indexes.
    """

    def test_consistency_with_ivfpq(self):
        """
        Compare IVFPQFastScan filtered results with IVFPQ (non-fastscan)
        filtered results. Both should only return allowed IDs.
        """
        d = 32
        ds = datasets.SyntheticDataset(d, 2000, 500, 20)

        rs = np.random.RandomState(42)
        subset = rs.choice(500, 150, replace=False).astype("int64")
        sel = faiss.IDSelectorBatch(subset)
        allowed = set(subset.tolist())

        # FastScan index
        index_fs = faiss.index_factory(d, "IVF32,PQ4x4fs")
        index_fs.train(ds.get_train())
        index_fs.add(ds.get_database())

        # Non-fastscan PQ index
        index_pq = faiss.index_factory(d, "IVF32,PQ4x4np")
        index_pq.train(ds.get_train())
        index_pq.add(ds.get_database())

        k = 10
        params = faiss.SearchParametersIVF(sel=sel, nprobe=8)

        Dfs, Ifs = index_fs.search(ds.get_queries(), k, params=params)
        Dpq, Ipq = index_pq.search(ds.get_queries(), k, params=params)

        # Both should only return allowed IDs
        for q in range(ds.nq):
            for j in range(k):
                if Ifs[q, j] >= 0:
                    self.assertIn(int(Ifs[q, j]), allowed)
                if Ipq[q, j] >= 0:
                    self.assertIn(int(Ipq[q, j]), allowed)

    def test_blockskip_consistency_with_ivfpq(self):
        """
        Compare IVFPQFastScan vs IVFPQ (non-fastscan) when using
        IDSelectorNot that excludes entire 32-vector blocks.
        This specifically validates the block-skip optimization
        against a baseline that doesn't have block-skip logic.
        """
        d = 32
        nb = 500
        ds = datasets.SyntheticDataset(d, 2000, nb, 20)

        # Exclude entire blocks: IDs 0-31 and 64-95
        excluded = np.concatenate([
            np.arange(0, 32, dtype="int64"),
            np.arange(64, 96, dtype="int64"),
        ])
        inner_sel = faiss.IDSelectorBatch(excluded)
        sel = faiss.IDSelectorNot(inner_sel)
        allowed = set(i for i in range(nb) if i not in excluded)

        # FastScan index
        index_fs = faiss.index_factory(d, "IVF32,PQ4x4fs")
        index_fs.train(ds.get_train())
        index_fs.add(ds.get_database())

        # Non-fastscan PQ index (no block-skip logic)
        index_pq = faiss.index_factory(d, "IVF32,PQ4x4np")
        index_pq.train(ds.get_train())
        index_pq.add(ds.get_database())

        k = 10
        params = faiss.SearchParametersIVF(sel=sel, nprobe=8)

        Dfs, Ifs = index_fs.search(ds.get_queries(), k, params=params)
        Dpq, Ipq = index_pq.search(ds.get_queries(), k, params=params)

        # Both should only return allowed IDs
        for q in range(ds.nq):
            for j in range(k):
                if Ifs[q, j] >= 0:
                    self.assertIn(int(Ifs[q, j]), allowed,
                        f"FastScan q={q} j={j}: id {Ifs[q, j]} is excluded")
                if Ipq[q, j] >= 0:
                    self.assertIn(int(Ipq[q, j]), allowed,
                        f"IVFPQ q={q} j={j}: id {Ipq[q, j]} is excluded")

    def test_partial_vs_whole_block_filter(self):
        """
        Excluding the same set of IDs via two identical IDSelectorNot
        configurations should produce identical results.
        """
        d = 32
        ds = datasets.SyntheticDataset(d, 2000, 200, 10)

        index = faiss.index_factory(d, "IVF32,PQ4x4fs")
        index.train(ds.get_train())
        index.add(ds.get_database())

        # Exclude the first 32 elements (whole block)
        excluded = np.arange(0, 32, dtype="int64")
        sel1 = faiss.IDSelectorNot(faiss.IDSelectorBatch(excluded))
        sel2 = faiss.IDSelectorNot(faiss.IDSelectorBatch(excluded))

        k = 10
        params1 = faiss.SearchParametersIVF(sel=sel1, nprobe=8)
        params2 = faiss.SearchParametersIVF(sel=sel2, nprobe=8)

        D1, I1 = index.search(ds.get_queries(), k, params=params1)
        D2, I2 = index.search(ds.get_queries(), k, params=params2)

        # Results should be identical
        np.testing.assert_array_equal(I1, I2)
        np.testing.assert_array_almost_equal(D1, D2, decimal=5)

    def test_heavy_filtering(self):
        """
        With >90% of vectors filtered, block-skip should skip most blocks.
        Verify correctness under heavy filtering.
        """
        d = 32
        ds = datasets.SyntheticDataset(d, 2000, 1000, 20)

        index = faiss.index_factory(d, "IVF32,PQ4x4fs")
        index.train(ds.get_train())
        index.add(ds.get_database())

        # Allow only ~5% of vectors
        rs = np.random.RandomState(99)
        subset = rs.choice(1000, 50, replace=False).astype("int64")
        sel = faiss.IDSelectorBatch(subset)
        allowed = set(subset.tolist())

        k = 10
        params = faiss.SearchParametersIVF(sel=sel, nprobe=16)
        Dfs, Ifs = index.search(ds.get_queries(), k, params=params)

        # All fastscan results must be in allowed set
        for q in range(ds.nq):
            for j in range(k):
                idx = int(Ifs[q, j])
                if idx >= 0:
                    self.assertIn(idx, allowed)


@for_all_simd_levels
class TestFastScanRangeSearchFilter(unittest.TestCase):
    """Test range_search with IDSelector on fastscan IVF indexes."""

    def test_range_search_filtered(self):
        """Range search with IDSelector should only return allowed IDs."""
        d = 32
        ds = datasets.SyntheticDataset(d, 2000, 500, 10)

        index = faiss.index_factory(d, "IVF32,PQ4x4fs")
        index.train(ds.get_train())
        index.add(ds.get_database())

        # Unfiltered range search to find a reasonable radius
        Dref, _ = index.search(ds.get_queries(), 10)
        radius = float(Dref.max()) * 1.5

        rs = np.random.RandomState(123)
        subset = rs.choice(500, 150, replace=False).astype("int64")
        sel = faiss.IDSelectorBatch(subset)
        allowed = set(subset.tolist())

        params = faiss.SearchParametersIVF(sel=sel, nprobe=8)

        try:
            lims, D, I = index.range_search(
                ds.get_queries(), radius, params=params
            )

            # All returned IDs should be in the allowed set
            for idx in I:
                if idx >= 0:
                    self.assertIn(
                        int(idx), allowed,
                        f"Range search returned id {idx} not in allowed set"
                    )
        except RuntimeError:
            # range_search may not be supported for all fastscan variants
            pass


if __name__ == "__main__":
    unittest.main()

File: tests/test_fastscan_filter.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <memory>
#include <unordered_set>
#include <vector>

#include <gtest/gtest.h>

#include <faiss/IndexFlat.h>
#include <faiss/IndexIVFPQFastScan.h>
#include <faiss/impl/IDSelector.h>
#include <faiss/index_factory.h>
#include <faiss/utils/random.h>

using namespace faiss;

namespace {

constexpr int d = 32;
constexpr size_t nt = 5000;
constexpr size_t nb = 1000;
constexpr size_t nq = 20;

std::vector<float> make_data(size_t n, int64_t seed) {
    std::vector<float> data(n * d);
    rand_smooth_vectors(n, d, data.data(), seed);
    return data;
}

/*************************************************************
 * Helper: test filtered search on a fastscan IVF index.
 *
 * Primary check: every returned ID is in the allowed set.
 * Secondary check: results have reasonable overlap with Flat
 * baseline (loose threshold since PQ is approximate).
 *************************************************************/

void test_filtered_search(
        const char* fs_index_key,
        MetricType metric,
        const size_t k,
        IDSelector& sel,
        const std::unordered_set<idx_t>& allowed_ids,
        const size_t db_size = nb) {
    auto xt = make_data(nt, 1234);
    auto xb = make_data(db_size, 4567);
    auto xq = make_data(nq, 7890);

    // Build fastscan index
    std::unique_ptr<Index> fs_index(index_factory(d, fs_index_key, metric));
    fs_index->train(nt, xt.data());
    fs_index->add(db_size, xb.data());

    // Search fastscan with selector
    SearchParametersIVF params;
    params.sel = &sel;
    params.nprobe = 8;

    std::vector<float> Dfs(nq * k);
    std::vector<idx_t> Ifs(nq * k);

    fs_index->search(nq, xq.data(), k, Dfs.data(), Ifs.data(), &params);

    // Primary check: all results are in the allowed set (or -1)
    for (size_t i = 0; i < nq * k; i++) {
        if (Ifs[i] >= 0) {
            EXPECT_TRUE(allowed_ids.count(Ifs[i]) > 0)
                    << "Fastscan returned id " << Ifs[i]
                    << " which is not in the allowed set";
        }
    }

    // Secondary check: if allowed set is large enough, we should get
    // some valid results (not all -1)
    if (allowed_ids.size() > k) {
        size_t valid_count = 0;
        for (size_t i = 0; i < nq * k; i++) {
            if (Ifs[i] >= 0)
                valid_count++;
        }
        EXPECT_GT(valid_count, 0) << "Expected some valid results";
    }
}

/*************************************************************
 * Test: IDSelectorBatch with fastscan (batch of allowed IDs)
 *************************************************************/

// Both tests are needed, because one tests CMax,
//   and another tests CMin.

TEST(TestFastScanFilter, IVFPQfs_Batch_L2) {
    std::vector<idx_t> subset;
    for (idx_t i = 0; i < nb; i += 3) { // ~333 out of 1000
        subset.push_back(i);
    }

    std::unordered_set<idx_t> allowed(subset.begin(), subset.end());
    IDSelectorBatch sel(subset.size(), subset.data());

    test_filtered_search("IVF32,PQ4x4fs", METRIC_L2, 10, sel, allowed);
}

TEST(TestFastScanFilter, IVFPQfs_Batch_IP) {
    std::vector<idx_t> subset;
    for (idx_t i = 0; i < nb; i += 3) {
        subset.push_back(i);
    }

    std::unordered_set<idx_t> allowed(subset.begin(), subset.end());
    IDSelectorBatch sel(subset.size(), subset.data());

    test_filtered_search(
            "IVF32,PQ4x4fs", METRIC_INNER_PRODUCT, 10, sel, allowed);
}

/*************************************************************
 * Test: k=1 (SingleResultHandler path) and k=40 (ReservoirHandler)
 *************************************************************/

TEST(TestFastScanFilter, IVFPQfs_K1) {
    std::vector<idx_t> subset;
    for (idx_t i = 0; i < nb; i += 2) {
        subset.push_back(i);
    }

    std::unordered_set<idx_t> allowed(subset.begin(), subset.end());
    IDSelectorBatch sel(subset.size(), subset.data());

    test_filtered_search("IVF32,PQ4x4fs", METRIC_L2, 1, sel, allowed);
}

TEST(TestFastScanFilter, IVFPQfs_K40) {
    std::vector<idx_t> subset;
    for (idx_t i = 0; i < nb; i += 2) {
        subset.push_back(i);
    }

    std::unordered_set<idx_t> allowed(subset.begin(), subset.end());
    IDSelectorBatch sel(subset.size(), subset.data());

    test_filtered_search("IVF32,PQ4x4fs", METRIC_L2, 40, sel, allowed);
}

/*************************************************************
 * Test: IDSelectorNot that excludes entire blocks (IDs 0-31, 64-95, etc.)
 * This specifically tests the block-skip optimization.
 *************************************************************/

TEST(TestFastScanFilter, BlockSkip_WholeBlocks) {
    // Exclude blocks 0 (ids 0-31) and 2 (ids 64-95)
    std::vector<idx_t> excluded;
    for (idx_t i = 0; i < 32; i++) {
        excluded.push_back(i);
    }
    for (idx_t i = 64; i < 96; i++) {
        excluded.push_back(i);
    }

    // Allowed = everything NOT in excluded
    std::unordered_set<idx_t> allowed;
    std::unordered_set<idx_t> excluded_set(excluded.begin(), excluded.end());
    for (idx_t i = 0; i < nb; i++) {
        if (excluded_set.count(i) == 0) {
            allowed.insert(i);
        }
    }

    IDSelectorBatch inner_sel(excluded.size(), excluded.data());
    IDSelectorNot sel(&inner_sel);

    test_filtered_search("IVF32,PQ4x4fs", METRIC_L2, 10, sel, allowed);
}

/*************************************************************
 * Test: Partial block filtering (exclude some IDs within a block)
 *************************************************************/

TEST(TestFastScanFilter, BlockSkip_PartialBlock) {
    // Exclude only a few IDs from the first block (0-31)
    std::vector<idx_t> excluded = {5, 10, 20, 31};
    std::unordered_set<idx_t> excluded_set(excluded.begin(), excluded.end());

    std::unordered_set<idx_t> allowed;
    for (idx_t i = 0; i < nb; i++) {
        if (excluded_set.count(i) == 0) {
            allowed.insert(i);
        }
    }

    IDSelectorBatch inner_sel(excluded.size(), excluded.data());
    IDSelectorNot sel(&inner_sel);

    test_filtered_search("IVF32,PQ4x4fs", METRIC_L2, 10, sel, allowed);
}

/*************************************************************
 * Test: Empty selector (nothing accepted) -> all results -1
 *************************************************************/

TEST(TestFastScanFilter, EmptySelector) {
    auto xt = make_data(nt, 1234);
    auto xb = make_data(nb, 4567);
    auto xq = make_data(nq, 7890);

    std::unique_ptr<Index> index(index_factory(d, "IVF32,PQ4x4fs", METRIC_L2));
    index->train(nt, xt.data());
    index->add(nb, xb.data());

    // Empty batch selector: accepts nothing
    std::vector<idx_t> empty_subset;
    IDSelectorBatch sel(0, empty_subset.data());

    size_t k = 10;
    std::vector<float> D(nq * k);
    std::vector<idx_t> I(nq * k);

    SearchParametersIVF params;
    params.sel = &sel;
    params.nprobe = 4;
    index->search(nq, xq.data(), k, D.data(), I.data(), &params);

    // All results should be -1 (nothing matches)
    for (size_t i = 0; i < nq * k; i++) {
        EXPECT_EQ(I[i], -1)
                << "Expected -1 at position " << i << " but got " << I[i];
    }
}

/*************************************************************
 * Test: ntotal not a multiple of 32
 * Tests the min(32, ntotal - j0) boundary in block-skip check.
 *************************************************************/

TEST(TestFastScanFilter, NonAlignedNtotal_50) {
    std::vector<idx_t> subset;
    for (idx_t i = 0; i < 50; i += 2) {
        subset.push_back(i);
    }

    std::unordered_set<idx_t> allowed(subset.begin(), subset.end());
    IDSelectorBatch sel(subset.size(), subset.data());

    test_filtered_search("IVF32,PQ4x4fs", METRIC_L2, 10, sel, allowed, 50);
}

TEST(TestFastScanFilter, NonAlignedNtotal_77) {
    std::vector<idx_t> subset;
    for (idx_t i = 0; i < 77; i += 2) {
        subset.push_back(i);
    }

    std::unordered_set<idx_t> allowed(subset.begin(), subset.end());
    IDSelectorBatch sel(subset.size(), subset.data());

    test_filtered_search("IVF32,PQ4x4fs", METRIC_L2, 10, sel, allowed, 77);
}

TEST(TestFastScanFilter, NonAlignedNtotal_150) {
    std::vector<idx_t> subset;
    for (idx_t i = 0; i < 150; i += 3) {
        subset.push_back(i);
    }

    std::unordered_set<idx_t> allowed(subset.begin(), subset.end());
    IDSelectorBatch sel(subset.size(), subset.data());

    test_filtered_search("IVF32,PQ4x4fs", METRIC_L2, 10, sel, allowed, 150);
}

/*************************************************************
 * Test: IDSelectorRange with fastscan
 *************************************************************/

TEST(TestFastScanFilter, IVFPQfs_Range) {
    std::unordered_set<idx_t> allowed;
    for (idx_t i = 100; i < 500; i++) {
        allowed.insert(i);
    }

    IDSelectorRange sel(100, 500);

    test_filtered_search("IVF32,PQ4x4fs", METRIC_L2, 10, sel, allowed);
}

/*************************************************************
 * Test: Heavy filtering (>90% excluded) - block-skip should
 * skip most blocks entirely.
 *************************************************************/

TEST(TestFastScanFilter, HeavyFiltering) {
    // Allow only ~5% of vectors
    std::vector<idx_t> subset;
    for (idx_t i = 0; i < nb; i += 20) {
        subset.push_back(i);
    }

    std::unordered_set<idx_t> allowed(subset.begin(), subset.end());
    IDSelectorBatch sel(subset.size(), subset.data());

    test_filtered_search("IVF32,PQ4x4fs", METRIC_L2, 10, sel, allowed);
}

} // namespace
